src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs

```diff
@@ -0,0 +1,110 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Options;
+using NuGet.Services;
+using NuGet.Versioning;
+
+namespace NuGet.Jobs.Catalog2Registration
+{
+    public class RegistrationUrlBuilder
+    {
+        private readonly IOptionsSnapshot<Catalog2RegistrationConfiguration> _options;
+        private readonly string _legacyBaseUrl;
+        private readonly string _gzippedBaseUrl;
+        private readonly string _semver2BaseUrl;
+
+        public RegistrationUrlBuilder(IOptionsSnapshot<Catalog2RegistrationConfiguration> options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+
+            _legacyBaseUrl = EnsureTrailingSlash(_options.Value.LegacyBaseUrl);
+            _gzippedBaseUrl = EnsureTrailingSlash(_options.Value.GzippedBaseUrl);
+            _semver2BaseUrl = EnsureTrailingSlash(_options.Value.SemVer2BaseUrl);
+        }
+
+        public string GetIndexPath(string id)
+        {
+            return $"{Uri.EscapeUriString(id.ToLowerInvariant())}/index.json";
+        }
+
+        public string GetIndexUrl(HiveType hive, string id)
+        {
+            return GetBaseUrl(hive) + GetIndexPath(id);
+        }
+
+        public string GetInlinedPageUrl(HiveType hive, string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return GetIndexUrl(hive, id) + "#" + GetPageFragment(lower, upper);
+        }
+
+        public string ConvertHive(HiveType fromHive, HiveType toHive, string url)
+        {
+            if (url == null)
+            {
+                return null;
+            }
+
+            var path = ConvertToPath(fromHive, url);
+            return GetBaseUrl(toHive) + path;
+        }
+
+        public string ConvertToPath(HiveType hive, string url)
+        {
+            var baseUrl = GetBaseUrl(hive);
+            Guard.Assert(url.StartsWith(baseUrl), $"URL '{url}' does not start with expected base URL '{baseUrl}'.");
+            return url.Substring(baseUrl.Length);
+        }
+
+        public string GetPageUrl(HiveType hive, string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return GetBaseUrl(hive) + GetPagePath(id, lower, upper);
+        }
+
+        public string GetPagePath(string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return $"{Uri.EscapeUriString(id.ToLowerInvariant())}/{GetPageFragment(lower, upper)}.json";
+        }
+
+        private static string GetPageFragment(NuGetVersion lower, NuGetVersion upper)
+        {
+            return
+                $"page/" +
+                $"{Uri.EscapeUriString(lower.ToNormalizedString().ToLowerInvariant())}/" +
+                $"{Uri.EscapeUriString(upper.ToNormalizedString().ToLowerInvariant())}";
+        }
+
+        public string GetLeafPath(string id, NuGetVersion version)
+        {
```

Comment:
Null checks?

Reply:
In general I don't put null checks are all the levels. This is a low level component so I thought we wouldn't need to null check here. If you feel strongly I can put them in.

Reply:
Even though low level, it's a public API (from technical point of view). 
I personally would just add them to the places I highlighted, as the other public methods are calling into these and would be implicitly covered. Your call :)

Reply:
Okay, sounds good. I'll make that change.

Reply:
Done.